pysparkling/sql/tests/test_types_parser.py

```diff
@@ -0,0 +1,256 @@
+import re
+
+import pytest
+
+from pysparkling.sql.parsers.types_parser import parser
+from pysparkling.sql.types import (
+    ArrayType, BinaryType, BooleanType, ByteType, DataType, DateType, DecimalType, DoubleType, FloatType, IntegerType,
+    LongType, MapType, NullType, ShortType, StringType, StructField, StructType, TimestampType
+)
+from pysparkling.sql.utils import ParseException
+
+# Tests copied from sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DataTypeParserSuite.scala
+# commit: c62b84a
+
+
+INVALID_CHARS = re.compile('[^_a-z0-9]*', flags=re.IGNORECASE)
+
+
+# Dynamic creation of test methods, much like done in Scala
+def checkDataType(dataTypeString: str, expectedDataType: DataType):
+    name = INVALID_CHARS.sub('', dataTypeString)
+
+    print(f"""
+def test_{name}():
+    assert parser.parse("{dataTypeString}") == {expectedDataType}()
+    """)
+
+
+def unsupported(dataTypeString: str):
+    name = INVALID_CHARS.sub('', dataTypeString)
+
+    print(f"""
+def test_{name}_is_not_supported():
+    with pytest.raises(ParseException):
+        parser.parse("{dataTypeString}")
+""")
+
+
+def test_int():
+    assert parser.parse("int") == IntegerType()
+
+
+def test_integer():
+    assert parser.parse("integer") == IntegerType()
+
+
+def test_BooLean():
+    assert parser.parse("BooLean") == BooleanType()
+
+
+def test_tinYint():
+    assert parser.parse("tinYint") == ByteType()
+
+
+def test_smallINT():
+    assert parser.parse("smallINT") == ShortType()
+
+
+def test_INT():
+    assert parser.parse("INT") == IntegerType()
+
+
+def test_INTEGER():
+    assert parser.parse("INTEGER") == IntegerType()
+
+
+def test_bigint():
+    assert parser.parse("bigint") == LongType()
+
+
+def test_float():
+    assert parser.parse("float") == FloatType()
+
+
+def test_dOUBle():
+    assert parser.parse("dOUBle") == DoubleType()
+
+
+def test_decimal105():
+    assert parser.parse("decimal(10, 5)") == DecimalType(10, 5)
+
+
+def test_decimal():
+    assert parser.parse("decimal") == DecimalType()
+
+
+def test_Dec105():
+    assert parser.parse("Dec(10, 5)") == DecimalType(10, 5)
+
+
+def test_deC():
+    assert parser.parse("deC") == DecimalType()
+
+
+def test_DATE():
+    assert parser.parse("DATE") == DateType()
+
+
+def test_timestamp():
+    assert parser.parse("timestamp") == TimestampType()
+
+
+def test_string():
+    assert parser.parse("string") == StringType()
+
+
+# checkDataType("ChaR(5)", CharType(5))
+# checkDataType("ChaRacter(5)", CharType(5))
+# checkDataType("varchAr(20)", VarcharType(20))
+# checkDataType("cHaR(27)", CharType(27))
+
+
+def test_BINARY():
+    assert parser.parse("BINARY") == BinaryType()
+
+
+def test_void():
+    assert parser.parse("void") == NullType()
```

Comment:
Where is this from?

```python
>>> spark.createDataFrame([], "id void")
[...[
pyspark.sql.utils.ParseException: 
DataType void is not supported.(line 1, pos 3)

== SQL ==
id void
---^^^
```

Reply:
https://github.com/apache/spark/blob/54c053afb0c9d3fcc7ac311100c8db9deeb163c0/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/parser/DataTypeParserSuite.scala#L64